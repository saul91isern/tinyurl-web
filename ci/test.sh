#!/bin/sh

echo "----- Install dependencies -----"
yarn
echo "----- Linter -----"
yarn test --ci

