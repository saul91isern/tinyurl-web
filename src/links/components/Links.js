import _ from "lodash/fp";
import React, { useState } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import {
  Dimmer,
  Icon,
  Input,
  Loader,
  Message,
  Popup,
  Table,
} from "semantic-ui-react";
import { deleteLink, setQuery } from "../routines";
import { getLinksChunk } from "../selectors";
import Pagination from "./Pagination";

const timeout = 1000;

const CopyCell = ({ iconId }) => {
  const [open, setOpen] = useState(false);
  const copy = (id) => {
    const text = document.getElementById(id).firstChild?.data;
    navigator.clipboard.writeText(text);

    !open && setOpen(true);

    setTimeout(() => {
      setOpen(false);
    }, timeout);
  };

  return (
    <Popup
      trigger={
        <Icon name={"copy"} link onClick={() => copy(`link_${iconId}`)} />
      }
      content={"Copied!"}
      on="click"
      open={open}
      position="top center"
    />
  );
};

CopyCell.propTypes = {
  iconId: PropTypes.number,
};

export const Links = ({ loading, deleteLink, linksQuery, links, setQuery }) => {
  const baseURL = process.env.REACT_APP_API_HOST;

  const onInputChange = (_e, { value }) => setQuery(value);

  return (
    <>
      <Input
        icon={{ name: "search", circular: true }}
        placeholder="Search..."
        onChange={onInputChange}
        value={linksQuery || ""}
      />
      {_.isEmpty(links) && !loading ? (
        <Message
          icon="question"
          header="No results found"
          content="Maybe you want shorten a new url first."
        />
      ) : (
        <>
          <Dimmer.Dimmable dimmed={loading}>
            <Dimmer active={loading} inverted>
              <Loader />
            </Dimmer>
            <Table striped collapsing>
              <Table.Header>
                <Table.Row>
                  <Table.HeaderCell>Complete Url</Table.HeaderCell>
                  <Table.HeaderCell>Shortened Url</Table.HeaderCell>
                  <Table.HeaderCell>Copy</Table.HeaderCell>
                  <Table.HeaderCell>Delete</Table.HeaderCell>
                </Table.Row>
              </Table.Header>
              <Table.Body>
                {links.map(({ url, hash }, key) => (
                  <Table.Row key={`row_${key}`}>
                    <Table.Cell>
                      <Popup
                        content={url}
                        trigger={
                          <a
                            href={url}
                            target="_blank"
                            rel="noopener noreferrer"
                          >
                            {_.truncate({ length: 100 })(url)}
                          </a>
                        }
                      />
                    </Table.Cell>
                    <Table.Cell>
                      <Popup
                        content={`${baseURL}/${hash}`}
                        trigger={
                          <a
                            id={`link_${key}`}
                            href={`${baseURL}/${hash}`}
                            target="_blank"
                            rel="noopener noreferrer"
                          >
                            {_.truncate({ length: 100 })(`${baseURL}/${hash}`)}
                          </a>
                        }
                      />
                    </Table.Cell>
                    <Table.Cell>
                      <CopyCell iconId={key} />
                    </Table.Cell>
                    <Table.Cell>
                      <Icon
                        name={"trash"}
                        link
                        onClick={() => deleteLink(hash)}
                      />
                    </Table.Cell>
                  </Table.Row>
                ))}
              </Table.Body>
            </Table>
            <Pagination />
          </Dimmer.Dimmable>
        </>
      )}
    </>
  );
};

Links.propTypes = {
  deleteLink: PropTypes.func,
  links: PropTypes.array,
  linksQuery: PropTypes.string,
  setQuery: PropTypes.func,
  loading: PropTypes.bool,
};

const mapStateToProps = (state) => ({
  linksQuery: state.linksQuery,
  links: getLinksChunk(state),
  loading: state.linksLoading,
});

const mapDispatchToProps = { deleteLink, setQuery };

export default connect(mapStateToProps, mapDispatchToProps)(Links);
