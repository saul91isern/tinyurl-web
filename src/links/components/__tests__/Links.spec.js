/* eslint-disable quotes */
import React from "react";
import { waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { render } from "../../../test/render";
import { Links } from "../Links";

describe("<Links />", () => {
  const loading = false;
  const deleteLink = jest.fn();
  const setQuery = jest.fn();
  const linksQuery = "";
  const links = [
    { url: "http://foo/bar", hash: "FOO" },
    { url: "http://baz/xyz", hash: "BAR" },
  ];

  it("renders links when not empty", async () => {
    const { container, queryByText } = render(
      <Links
        loading={loading}
        deleteLink={deleteLink}
        links={links}
        setQuery={setQuery}
        linksQuery={linksQuery}
      />,
      {}
    );
    expect(queryByText(/http:\/\/foo\/bar/)).toBeTruthy();
    expect(queryByText(/http:\/\/baz\/xyz/)).toBeTruthy();
    expect(queryByText(/FOO/)).toBeTruthy();
    expect(queryByText(/BAR/)).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it("renders message on empty links", async () => {
    const { container, queryByText } = render(
      <Links
        loading={loading}
        deleteLink={deleteLink}
        links={[]}
        setQuery={setQuery}
        linksQuery={linksQuery}
      />,
      {}
    );
    expect(queryByText(/http:\/\/foo\/bar/)).toBeFalsy();
    expect(queryByText(/http:\/\/baz\/xyz/)).toBeFalsy();
    expect(queryByText(/FOO/)).toBeFalsy();
    expect(queryByText(/BAR/)).toBeFalsy();
    expect(queryByText(/Maybe you want shorten a new url first./)).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it("renders Loader if loading", async () => {
    const { container } = render(
      <Links
        loading={true}
        deleteLink={deleteLink}
        links={[]}
        setQuery={setQuery}
        linksQuery={linksQuery}
      />,
      {}
    );
    expect(container.querySelector('[class="ui loader"]')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it("sets query on input change", async () => {
    const { findByPlaceholderText } = render(
      <Links
        loading={false}
        deleteLink={deleteLink}
        links={links}
        setQuery={setQuery}
        linksQuery={linksQuery}
      />,
      {}
    );
    userEvent.type(await findByPlaceholderText(/Search/), "q");
    await waitFor(() => expect(setQuery).toHaveBeenCalledWith("q"));
  });

  it("calls deleteLink on icon delete click", async () => {
    const { container } = render(
      <Links
        loading={false}
        deleteLink={deleteLink}
        links={links}
        setQuery={setQuery}
        linksQuery={linksQuery}
      />,
      {}
    );
    userEvent.click(await container.querySelector('[class="trash link icon"]'));
    expect(deleteLink).toHaveBeenCalledWith(links[0]?.hash);
  });
});
