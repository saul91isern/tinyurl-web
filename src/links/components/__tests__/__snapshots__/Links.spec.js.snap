// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Links /> renders Loader if loading 1`] = `
<div>
  <div
    class="ui icon input"
  >
    <input
      placeholder="Search..."
      type="text"
      value=""
    />
    <i
      aria-hidden="true"
      class="search circular icon"
    />
  </div>
  <div
    class="dimmed dimmable"
  >
    <div
      class="ui active transition visible inverted dimmer"
      style="display: flex;"
    >
      <div
        class="content"
      >
        <div
          class="ui loader"
        />
      </div>
    </div>
    <table
      class="ui collapsing striped table"
    >
      <thead
        class=""
      >
        <tr
          class=""
        >
          <th
            class=""
          >
            Complete Url
          </th>
          <th
            class=""
          >
            Shortened Url
          </th>
          <th
            class=""
          >
            Copy
          </th>
          <th
            class=""
          >
            Delete
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      />
    </table>
    Pagination
  </div>
</div>
`;

exports[`<Links /> renders links when not empty 1`] = `
<div>
  <div
    class="ui icon input"
  >
    <input
      placeholder="Search..."
      type="text"
      value=""
    />
    <i
      aria-hidden="true"
      class="search circular icon"
    />
  </div>
  <div
    class="dimmable"
  >
    <div
      class="ui inverted dimmer"
    >
      <div
        class="content"
      >
        <div
          class="ui loader"
        />
      </div>
    </div>
    <table
      class="ui collapsing striped table"
    >
      <thead
        class=""
      >
        <tr
          class=""
        >
          <th
            class=""
          >
            Complete Url
          </th>
          <th
            class=""
          >
            Shortened Url
          </th>
          <th
            class=""
          >
            Copy
          </th>
          <th
            class=""
          >
            Delete
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      >
        <tr
          class=""
        >
          <td
            class=""
          >
            <a
              href="http://foo/bar"
              rel="noopener noreferrer"
              target="_blank"
            >
              http://foo/bar
            </a>
          </td>
          <td
            class=""
          >
            <a
              href="http://localhost:4000/FOO"
              id="link_0"
              rel="noopener noreferrer"
              target="_blank"
            >
              http://localhost:4000/FOO
            </a>
          </td>
          <td
            class=""
          >
            <i
              aria-hidden="true"
              class="copy link icon"
            />
          </td>
          <td
            class=""
          >
            <i
              aria-hidden="true"
              class="trash link icon"
            />
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class=""
          >
            <a
              href="http://baz/xyz"
              rel="noopener noreferrer"
              target="_blank"
            >
              http://baz/xyz
            </a>
          </td>
          <td
            class=""
          >
            <a
              href="http://localhost:4000/BAR"
              id="link_1"
              rel="noopener noreferrer"
              target="_blank"
            >
              http://localhost:4000/BAR
            </a>
          </td>
          <td
            class=""
          >
            <i
              aria-hidden="true"
              class="copy link icon"
            />
          </td>
          <td
            class=""
          >
            <i
              aria-hidden="true"
              class="trash link icon"
            />
          </td>
        </tr>
      </tbody>
    </table>
    Pagination
  </div>
</div>
`;

exports[`<Links /> renders message on empty links 1`] = `
<div>
  <div
    class="ui icon input"
  >
    <input
      placeholder="Search..."
      type="text"
      value=""
    />
    <i
      aria-hidden="true"
      class="search circular icon"
    />
  </div>
  <div
    class="ui icon message"
  >
    <i
      aria-hidden="true"
      class="question icon"
    />
    <div
      class="content"
    >
      <div
        class="header"
      >
        No results found
      </div>
      <p>
        Maybe you want shorten a new url first.
      </p>
    </div>
  </div>
</div>
`;
